producer/writer/consumer_writer.go

```diff
@@ -0,0 +1,293 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package writer
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3msg/generated/proto/msgpb"
+	"github.com/m3db/m3msg/protocol/proto"
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+
+	"github.com/uber-go/tally"
+	"go.uber.org/atomic"
+)
+
+const (
+	defaultRetryForever = true
+)
+
+var (
+	defaultConn = new(net.TCPConn)
+)
+
+type consumerWriter interface {
+	// Write writes the marshaler, it should be thread safe.
+	Write(m proto.Marshaler) error
+
+	// Init initializes the consumer writer.
+	Init()
+
+	// Close closes the consumer writer.
+	Close()
+}
+
+type consumerWriterMetrics struct {
+	ackError          tally.Counter
+	decodeError       tally.Counter
+	encodeError       tally.Counter
+	resetConn         tally.Counter
+	resetError        tally.Counter
+	connectError      tally.Counter
+	setKeepAliveError tally.Counter
+}
+
+func newConsumerWriterMetrics(scope tally.Scope) consumerWriterMetrics {
+	return consumerWriterMetrics{
+		ackError:          scope.Counter("ack-error"),
+		decodeError:       scope.Counter("decode-error"),
+		encodeError:       scope.Counter("encode-error"),
+		resetConn:         scope.Counter("reset-conn"),
+		resetError:        scope.Counter("reset-conn-error"),
+		connectError:      scope.Counter("connect-error"),
+		setKeepAliveError: scope.Counter("set-keep-alive-error"),
+	}
+}
+
+type connectFn func(addr string) (net.Conn, error)
+
+type consumerWriterImpl struct {
+	// encodeLock controls the access to the encode function.
```

Comment:
Why not have `encdec` be responsible for locking the encode and decode functions? Seems odd that the caller has to be aware to lock different methods on a single interface.

Reply:
Because the `EncodeDecoder` can be used in cases where it is only accessed in single thread, so I decided to push the lock to the users.

Reply:
Since it's an interface, would it make sense to have implementations for both cases? The multithreaded one could be a small wrapper around the current one perhaps?